### ISSUE TITLE
Adding mrs_uav_trackers to existing projects

### ISSUE THREAD

**Issue by dookei** [ctu-mrs/mrs_uav_trackers#5]
Hi there,
I have currently an working project that I have been working on it for some time. I added already some functionalities and I would like now to try some other things like pure pursuit / path following. This seems to be well developed and tested on this package. Which I would love to test on my system.
Is there an "easy" way to integrate this package in existing "non-mrs" projects? I mean is there an generic interface where I could provide my pose/odom input and get a pose output for the offboard controller? 
Many thanks and great work guys!

**Comment by klaxalk** [comment]
Hey, with a little bit of work and refactoring I think you can use mrs_uav_trackers outside of mrs system. I assume that you are most interested in the MpcTracker. The trackers within this package follow this interface (https://github.com/ctu-mrs/mrs_uav_managers/blob/master/include/mrs_uav_managers/tracker.h)[https://github.com/ctu-mrs/mrs_uav_managers/blob/master/include/mrs_uav_managers/tracker.h]. So if you will emulate the actions of the ControlManager, the code will work for you. Which means:

- call the `initialize()` method first.
- call the `activate()` method, this starts timers within the tracker and sets up the initial conditions for the inner states
- cal the `update()` method repeatedly to get the current reference for your offboard controller
- call `setReference()`, setTrajectoryReference(), ..., on demand to set the references

There is also stuff that the tracker needs from the ControlManager, mainly this (https://github.com/ctu-mrs/mrs_uav_managers/blob/master/include/mrs_uav_managers/common_handlers.h)[https://github.com/ctu-mrs/mrs_uav_managers/blob/master/include/mrs_uav_managers/common_handlers.h) but I would just cut this out being you. So for the beginning, I would make all the function implementations to "be happy" and return true to not cause problems.

**Comment by dookei** [comment]
Hi @klaxalk, thanks for your prompt reply and sorry for my late reply ;-) .Yes, I am most definitely interested in the MPC tracker. I have my system currently up and running with V-SLAM, CAS and some other stuff all in off-board mode. I can load a set of waypoints and "very very rudimentary" go from one to the next. But as you might imagine this is quite primitive. I need a smoother and reliable controlled approach. Therefor the MPC tracker. Like you wrote this is the workhorse :-)! 
Quick questions:
How is the interface with the PX4/ UAV? Set position? Velocity ? acceleration? 
The  libMpcTrackerSolver.so is closed source? 
Do I need the https://github.com/ctu-mrs/mrs_uav_trajectory_generation in order to run this? Or can I just provide the setpoint with my existing node? 


I will be performing the integration/adaptation from your project to our system at the moment. I will keep you posted.
Keep up the good work!

**Comment by klaxalk** [comment]
Hey @dookei. So you could technically use the MPC tracker as you suggest, meaning for supplying a reference to PX4. In that case, the reference should definitely be the desired position, and its derivatives, the more the better (you can get smooth derivatives up to jerk with the tracker). However, I am not sure how this form of supplying reference works with PX4, we have never tried it. Instead, we do our own position+velocity control together with our own attitude control. See our [paper](https://arxiv.org/abs/2008.08050) for more info. Therefore, we only supply the attitude rate + thrust reference to PX4.

The libMpcTrackerSolver.so as well as the libMpcControllerSolver.so are pre-compiled for two reasons. First, they take approx. 10 minutes to compile and since rarely anything changes in them (they are just solver wrappers), we just have those `.so`s prepared in advance. Moreover, we are not permitted to share the sources of the solver we use.

You don't need the [mrs_uav_trajectory_generation](https://github.com/ctu-mrs/mrs_uav_trajectory_generation) package for single setpoints. The MPC tracker will handle setpoint references just fine (via the `setReference()` routine). But where the MPC tracker shines is with dynamic trajectories. Those can be however created in many ways, the  [mrs_uav_trajectory_generation](https://github.com/ctu-mrs/mrs_uav_trajectory_generation) package is just one of them.

Thanks, let me know how your work goes.

**Comment by dookei** [comment]
Hi @klaxalk , thanks for the previous messages. 
Currently I am just going through all the code to understand it and adapt it. I dunno if position or velocity instead of attitude  control is better but to me it sounds more easy to start from. And my goal was to leave the final stabilization and attitude control to the Px4.

Have you ever bench marked the inear mrs-MPC   tracker against others ? Like the one described by ETH Zürich  [Linear vs Nonlinear MPC for Trajectory Tracking](https://www.sciencedirect.com/science/article/pii/S2405896317313083)? According to the conclusions this seems to have a slightly better performance.

**Comment by dookei** [comment]
Hi @klaxalk, I believe I managed to port successfully the mrs\_uav\_trackers/ mpc\_tracker.cpp with only dependencies being on the  packages mrs\_lib and mrs_msg.

Under the `initialize()` I left the following out. Hope this not needed. Or ?

[https://github.com/ctu-mrs/mrs\_uav\_trackers/blob/ab0dc6698ba5735abbfcc79e2b079525cdc15bd2/src/mpc\_tracker/mpc\_tracker.cpp#L396](https://github.com/ctu-mrs/mrs_uav_trackers/blob/ab0dc6698ba5735abbfcc79e2b079525cdc15bd2/src/mpc_tracker/mpc_tracker.cpp#L396)

[https://github.com/ctu-mrs/mrs\_uav\_trackers/blob/ab0dc6698ba5735abbfcc79e2b079525cdc15bd2/src/mpc\_tracker/mpc\_tracker.cpp#L570](https://github.com/ctu-mrs/mrs_uav_trackers/blob/ab0dc6698ba5735abbfcc79e2b079525cdc15bd2/src/mpc_tracker/mpc_tracker.cpp#L570)

It loads the mpc_tracker.yaml parameters, creates the timers etc etc etc . So far so good.

I am following the interface [https://github.com/ctu-mrs/mrs\_uav\_managers/blob/master/include/mrs\_uav\_managers/tracker.h](https://github.com/ctu-mrs/mrs_uav_managers/blob/master/include/mrs_uav_managers/tracker.h)  like you recommended. 

But now I am having some difficulties under the `activate (const mrs_msgs::PositionCommand::ConstPtr& last_position_cmd)`.

Perhaps you could explain me better how is the interface to this mpc_tracker made? What do I need to provide what comes out and in what format? Black-Box style ;-)

I am following the control_manager.cpp in order to try to produce the same results.

My objective is to make this package portable( independent from mrs\_uav\_controller/manager/ and ) to other projects. So that basically you just copy it and provide the parameters, reference trajectory, inputs, etc and "magically" get the trajectory to track. I know is a little more complex  than that. But that is my  goal at the moment.

Many thanks!

**Comment by klaxalk** [comment]
Hey, great qestions, we will make it work, that should not be a problem. I will get back to you on monday, I am away of pc currently.

**Comment by dookei** [comment]
Great many Thanks!

**Comment by klaxalk** [comment]
Hey, sorry for the delay. I am not sure if I fully understand your environment and if you fully understand what the MPC tracker does.

- Is your package in ROS? If yes, then you don't have to leave out the `param_loader`. You can, if you want, load the parameters the normal way, i.e. `nh_.load_params(...)`, but I would just copy the param loader's [header](https://github.com/ctu-mrs/mrs_lib/blob/master/include/mrs_lib/param_loader.h) into your package. The same could be done with the message definition which is baked into the tracker. Those are the PositionCommand, Reference, ReferenceStamped, TrackerTrajectory, etc. You can just copy the message definitions from the `mrs_msgs` package and build them within your package. However, if you don't use ROS at all, you will need to load the params somehow manually (maybe using some c++ yaml library). Regarding the message objects, you can create custom classes that will contain the same data as the message classes, such that the tracker code does not need to change much.
- the `activate()` method is supposed to be called just before the tracker is meant to be put to work. The argument of the method is the last command produced by a previously active tracker. The command is passed to the next active tracker, so the tracker can set its internal state into the same state as the last tracker left it. This is to ensure that the generated reference will be smooth even when trackers switch. The tracker state is set to the current odometry If the command is unavailable. If you are not going to implement a tracker switching mechanism, just make sure your initial condition (the internal state set in the `activate()` method is set to whatever the true state of the UAV is.
- You write "*reference trajectory, inputs, etc and "magically" get the trajectory to track*". The tracker will not give you a trajectory to track, not in a complete form. It will, however, generate a reference in real-time for a controller. As time goes this will in fact draw a trajectory that resembles the desired trajectory and is feasible in all states. If you want to generate whole trajectories beforehand (which can be tracker by this tracker), you can use, e.g., the [trajectory generator](https://github.com/ctu-mrs/mrs_uav_trajectory_generation). You could use the MPC tracker to generate trajectories beforehand, but you would need to change the internal loop from the *real-time* varianto into some "pre-computation" variant that would run without delays (not in timers). You would then record the internal state until it reaches the desired end. But the intended cascade is the following:
1. Create a desired trajectory (not necessarily feasible), e.g., by [this tool](https://github.com/ctu-mrs/mrs_uav_trajectory_generation)
2. Give the trajectory (one-time event, or repeatedly if it changes in real time) to the MPC Tracker.
3. The MPC tracker follows the trajectory in real-time in the least-square position error manner while satisfying state constraints. The result is 100 Hz real-time all-state reference for controllers.
4. The controllers use the reference for feedback and feed forward to move the vehicle.

**Comment by dookei** [comment]
Hi Tomas,

many thanks for your detailed reply. Perhaps I will start to explain my case scenario. I started developing a "Framework" a "couple" months ago that aims indoor delivery drones.

The development and simulation environment is done under ROS(noetic) with Gazebo with Visual Studio Code(I like the debugging/customization/json interface on this). The target(UAV, also ROS) is a Nvidia Jetson with a Px4-FC connected over 2 serials(will change to ethernet/CAN later on).

At the moment I have one package(w/GUI) that manages the path(load, add/remove more  waypoints etc) during run time.  My goal now is to integrate the MRS-tracker (the MPC tracker is the best choice here, I assume) that feeds the setpoint dynamically to the Px4 (either position or velocity, I would like to leave the attitude control to the Px4). In between I have another middle layer(packages), that make the Interface with the PX4 convention and the trackers(landing/liftoff, MPC, etc) .  Some checks, mainly safety, are performed while feeding the dynamic-setpoint.

* * *

### What is  running so far.

I was able to port the [mpc_tracker.cpp](https://github.com/ctu-mrs/mrs_uav_trackers/blob/master/src/mpc_tracker/mpc_tracker.cpp) and modify some parts in order to load the parameters(that are loaded from the control\_manager.cpp) I am not sure what is needed at the moment so I just loaded all. From the MPC\_tracker package I can run already:

- [x] `Path_MPC_tracker.initialize(nh, "test_UAV", Path_MPC_tracker.mpc_CommonHandlers );`
    
- [x] `Path_MPC_tracker.loadConstraints(nh, Path_MPC_tracker.mpc_constraints);`
    
- [x] `Path_MPC_tracker.setConstraints(mrs_msgs::DynamicsConstraints::ConstPtr(std::make_unique<mrs_msgs::DynamicsConstraints>(Path_MPC_tracker.mpc_constraints)));`
    
- [ ] `Path_MPC_tracker.activate(last_position_cmd_);`
    

When I call activate()the timers start to run and I get(as expected) and exception at:

[https://github.com/ctu-mrs/mrs\_lib/blob/9778a1a03fd8ab6a28d76373df228e0a4653032f/src/attitude\_converter/attitude_converter.cpp#L459](https://github.com/ctu-mrs/mrs_lib/blob/9778a1a03fd8ab6a28d76373df228e0a4653032f/src/attitude_converter/attitude_converter.cpp#L459)

There still a lot of things that I am missing at the moment, until I get this working.

What I am currently still going through:

- [ ] `switchOdometrySource()`
    
- [ ] `setReference()`
    
- [ ] `setTrajectoryReference()`
    
- [ ] `startTrajectoryTracking()`
    
- [ ] `stopTrajectoryTracking()`
    
- [ ] `resumeTrajectoryTracking()`
    
- [ ] `enableCallbacks`
    
- [x] `hover()` this will be performed in my other package
    
- [ ] `getStatus`
    
- [ ] `update`
    

* * *

### Open Questions:

I have a lot of open questions at the moment as you might understand. Some I answer my self (sometimes while writing here) or while going through the code.

- Update()
    - The Odometry-input is feed on the update(const  mrs_msgs::UavState::ConstPtr  &  uav_state) right?  geometry_msgs/Pose ? is enough?
    - From the return. What could I use? Position? or velocity ? To be honest here I am kinda lost. 
- For later but nice to have 
- Do you believe it is possible to add  surrounding collision avoidance (when having a point-cloud, octomap)
- I also have a realtime feed from the trajectory from the other UAVs that could be used for UAV collision avoidance. I believe you have already this running right?
- Is this the right order to start the MPC Trackers ?
    1.  initialize
    2.  loadConstraints
    3.  setConstraints
    4.  setTrajectoryReference
    5.  setReference ... not sure here
    6.  startTrajectoryTracking
    7.  activate

Greetings,

Pedro

**Comment by klaxalk** [comment]
Hey, sorry for the delayed answer. I see you have been updating your reply, which is good. Regarding the open question:

- The input to the `update()` method is the UavState with all available information on the current state of the UAV, inc. speed, acceleration, etc. The `update()` method just saves the information and potentially uses it to set up initial conditions. You can substitute it with just the Pose if you wish.
- The return value of the `update()` is the **most important thing** about a tracker. It contains the desired position, acceleration, velocity, snap, heading, heading, rate, etc.  This reference is given to feedback controllers (e.g., the [SE(3) geometric feedback](https://github.com/ctu-mrs/mrs_uav_controllers/blob/master/src/se3_controller/se3_controller.cpp)). Some of it (acceleration, jerk) is used to calculate feed-forward control actions, which is a very important aspect of controlling a UAV. However, your feedback controller might not utilize all of the states in the reference. That is up to you. The MPC tracker makes sure all the states in the reference follow the model and are within the dynamics constraints.

The order of the routines should be the following:
1. `initialize()`
2. `setConstraints()` (however, you can just set the right constraints in the trackers yaml file, this `setConstraints()` routine is here for updating the constraints dynamically during the flight which you might not need)
2. `activate()`
3. !! now the tracker is active and producing a reference, so the `update()` is called repeatedly and its return value is supposed to be used !!.. no you can give the tracker commands, such as `setReference()` (rough goto coordinates), `setTrajectoryReference()` (sets time-parametrized trajectory), `hover()` (smoothly stops current motion)
4. `deactivate()` before you switch to another tracker

Adding a collision avoidance is outside of the scope of this tracker. This tracker just produces a feasible reference at a high rate for controllers with limited (and potentially infeasible) reference trajectories from a user. If you want to avoid obstacles,  design your desired trajectory (the input to the tracker) accordingly. The MPC Tracker will make sure the UAV's controllers are getting a feasible reference to follow your trajectory.

The mutual collision avoidance between the UAV's can be realized by the tracker, as described in [the publication](http://mrs.felk.cvut.cz/data/papers/iros_2018_mpc.pdf). It is meant for outside where the altitude flight level is not limited (the UAV's avoid by climbing higher). The tracker already subscribes to topics of other UAVs containing the future prediction of the UAV motion. I admit the code for the avoidance is far from nice and far from elegant, but it works. The trajectories are supposed to be transmitted in GPS Latitude+Longitude coordinates to avoid any confusion regarding coordinate frames (each UAV can potentially fly using a different localization system and having a different origin).

**Comment by dookei** [comment]
Hi Tomáš,
many thanks for you reply. I managed yesteday to make it work on my setup! 
There are a few things that I still need to change to fit my needs. At the moment I just provided a few waypoints to check the predicted trajectory and visualize it under Rviz. Looks exactly like what I need. 
I didnt test yet with dynamic odometry input. But I assume it should work without much hassle.

I woul like to thank you once again for your detailed replies. They helped me tremendously to understand how it works and where to look for. 
Many thanks! and keep up the good work!
